### PR TITLE
Refactor PodRunner and KubeTask to use PodController under the hood.

### PR DESCRIPTION
**Pull request by e-sumin**
## Change Overview

Here is a PodRunner refactoring using previously added PodController and PodCommandExecutor.

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [x] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Test Plan

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

**Comment by denisvmedia** [comment]
@e-sumin could you please fix the linter issues?

File: pkg/kube/pod_runner.go
Comment (denisvmedia) [inline pkg/kube/pod_runner.go:76]: It looks like this function is mostly copy-paste for Run. Can we reduce code duplication?

Comment (e-sumin) [reply pkg/kube/pod_runner.go:76]: the problem with the Run is that even though it already uses a PodController, it still calls a functor with the pod's name, and the functor works with the pod directly, not through the pod controller.
the RunEx was an attempt to move all users of the function step by step to use under the controller.

Comment (denisvmedia) [reply pkg/kube/pod_runner.go:76]: But I still see that they share 99% of each other. Can we reduce code duplication?

Comment (e-sumin) [reply pkg/kube/pod_runner.go:59]: Done.

File: pkg/kube/pod_runner.go
Comment (denisvmedia) [inline pkg/kube/pod_runner.go:54]: Nit. `ctx2` looks weird to me. I'd rename it to something more meaningful, e.g.:

```suggestion
	return p.RunEx(ctx, func(innerCtx context.Context, pc PodController) (PodOutputMap, error) {
		return fn(innerCtx, pc.Pod())
```

File: pkg/kube/pod_runner.go
Comment (denisvmedia) [inline pkg/kube/pod_runner.go:36]: Nit. I think we should either make named function types for both - Run and RunEx, or for none. The same is for the returned map type.

**Comment by viveksinghggits** [comment]
@e-sumin,
There are some other kanister functions (for example `prepare_data.go`) that use podRunner to create and execute pods in them. Are we planning to handle those in other/separate PRs?

File: pkg/kube/pod_runner.go
Comment (viveksinghggits) [reply pkg/kube/pod_runner.go:36]: I think we should add a comment where on why we have to methods that are doing to the same things. Basically what you have explained in the comment below.
Also, I think `RunEx` is not very intuitive, do we have other alternate names :) ?

**Review by viveksinghggits — COMMENTED** [review]
LGTM overall.

File: pkg/kube/pod_runner.go
Comment (e-sumin) [reply pkg/kube/pod_runner.go:41]: I've added the comment, but don't have better name. If you have something in mind, please suggest.

Comment (e-sumin) [reply pkg/kube/pod_runner.go:41]: @denisvmedia I made returned type similar for both functions, but was unable to invent name for functor type, so I left it as is.

**Comment by e-sumin** [comment]
@viveksinghggits yep, there are several places where pod runner is used. I think it's better to refactor them in separate PRs, so that the reviewing process will be faster.

File: pkg/kube/pod_runner.go
Comment (denisvmedia) [reply pkg/kube/pod_runner.go:41]: If no one has better names I suggest these:

- `PodAwareRunFunc`
- `PodControllerAwareRunFunc`

**Comment by denisvmedia** [comment]
@e-sumin please fix the signatures in the interface before merging. I put `hold off merging` to prevent the merge. You can remove it as soon as you sort out the issue.

File: pkg/kube/pod_runner.go
Comment (viveksinghggits) [reply pkg/kube/pod_runner.go:41]: I am assuming that we are going to get rid of `Run` method eventually. And then we can rename `RunEx` to `Run`.
I am ok with `RunEx` as well.

**Review by viveksinghggits — APPROVED** [review]
Approving, we can merge after addressing Denis' interface signature comment.

**Pull request merged**
